### DISCUSSION TITLE
1:1 Relationship circular dependency. How to create entities transactionally that reference each other.

### DISCUSSION THREAD

**Discussion by kwanhlee** [mikro-orm/mikro-orm#6131] [Q&A]
I have User and UserSettings Table. They are 1:1 relationship. However, since they are a circular relationship, it won't let me create both entities in a transaction. How can I get both entities create during the creation of User?

```
export abstract class BaseAttribute {
  @PrimaryKey()
  id: string = uuidv4().toString();

  @Property()
  created_at: Date = new Date();

  @Property({ onUpdate: () => new Date() })
  updated_at: Date = new Date();
}

export const baseAttributes = {
    id: { type: 'String', primary: true },
    created_at: { type: 'Date' },
    updated_at: { type: 'Date' }
};
```
```
@Entity()
export class User extends BaseAttribute {

  @Property()
  email: string;

  @OneToOne(() => UserSettings, userSettings => userSettings.user)
  userSettings?: UserSettings;
}

export const UserSchema = new EntitySchema<User>({
  name: 'User',
  tableName: 'user',
  properties: {
    ...baseAttributes,
    email: { type: 'string', unique: true },
    userSettings: { 
      kind: '1:1', 
      entity: () => 'UserSettings', 
      inversedBy: 'user',
      owner: true,
      cascade: [Cascade.REMOVE],
      nullable: true
    },  
  },
});
```

```
@Entity()
export class UserSettings extends BaseAttribute {

  @OneToOne(() => User, user => user.userSettings)
  user!: Rel<User>
}
  
export const UserSettingsSchema = new EntitySchema<UserSettings>({
    name: 'UserSettings',
    tableName: 'user_settings',
    properties: {
      ...baseAttributes,
      user: { 
        kind: '1:1', 
        entity: () => 'User', 
        mappedBy:'userSettings'
      },
    },
  });

```

Service function:
```
async createUser(
    body: registerUserField, 
  ): Promise<User> {
    return this.em.transactional(async (em) => {
      // Create User
      const userRepo = em.getRepository(UserSchema)
      const userSettingsRepo = em.getRepository(UserSettingsSchema)
    
  
      const user = new User();
      user.email = body.email;
      
    
      // Create UserSettings
      const userSettings = new UserSettings();
      userSettings.user = user;
      user.userSettings = userSettings; // setting this I think is the problem?
 
      const userSettingsEntity = userSettingsRepo.create({...userSettings})
      const userEntity = userRepo.create({...user})
     
      await em.persistAndFlush([userEntity, userSettingsEntity]);
      return user;
    });
  }
```
Running the above create user code gives me the error:
```
insert into "user" ("id", "created_at", "updated_at", "email", "user_settings_id") values ('58c0e7da-6dd4-4205-b230-558c4af3c7f6', '2024-10-12T07:01:39.964Z', '2024-10-12T07:01:39.964Z', 'test7@test.com', '[object Object]') - insert or update on table "user" violates foreign key constraint "user_user_settings_id_foreign"
 - detail: Key (user_settings_id)=([object Object]) is not present in table "user_settings".
```

_Originally posted by @kwanhlee in https://github.com/mikro-orm/mikro-orm/discussions/5776#discussioncomment-10921160_

**Comment by kwanhlee** [comment]
Also tried adding Rel<>:
https://github.com/OnetoOne(() => UserSettings, userSettings => userSettings.user)
userSettings?: Rel;

**Comment by B4nan** [comment]
Please don't do this. You just asked a question in the comments, and within 3 minutes, you moved it to an issue. Just wait (more than 3 minutes, lol), you will get your answer. If you want to open an issue, provide a complete reproduction, not snippets like this, and follow the issue template. If you are lost the way you clearly are, its better to come to discord or use GH discussions. You will only make things worse if you spam like this, it's not helpful.

I am not even sure what you are trying to show here, the snippets are all mixing decorators with EntitySchema which makes no sense, you even combine class based approach (with `class` option) with interfaces (with `name` option), those are exclusive, you got so many mistakes in the entity definitions that I am not sure where did you come up with those things, surely not from our docs. You basically mix 3 approaches in your entity definition.

Circular relations are supported, your problem (outside of that crazy mixture of entity definition approaches) is most likely the object desctructing - you also mix approaches there, creating entity instances via constructor and then via `em.create` - those things are doing the same!

```ts
const user = new User();
user.email = body.email;

// Create UserSettings
const userSettings = new UserSettings();
userSettings.user = user;
user.userSettings = userSettings;

await em.persistAndFlush([user, userSettings]);
return user;
```

**Reply by kwanhlee** [reply]
Understand. Thanks for the help.
